### PR TITLE
docker-for-mac/osxfs: clarify the ownership behavior in Beta 29

### PR DESCRIPTION
**Pull request by dsheets**
### Project version

Docker for Mac Beta 29

### Related issue or PR in another project

docker/pinata#5671

### CODE REVIEW

File: docker-for-mac/osxfs.md [hunk 1]
```diff
@@ -85,9 +85,11 @@ requests for ownership metadata will return the previously set
 values. Ownership-based permissions are only enforced at the OS X file
```
Comment (mstanleyjones) [inline docker-for-mac/osxfs.md]: s/OS X/macOS
s/file system/filesystem

File: docker-for-mac/osxfs.md [hunk 2]
```diff
@@ -85,9 +85,11 @@ requests for ownership metadata will return the previously set
...
 system level with all accessing processes behaving as the user running
 Docker. If the user does not have permission to read extended attributes
-on an object, e.g. when that object's permissions are `0000`, ownership
-will be reported as the accessing process until the extended attribute
-is again readable.
+on an object, e.g. when that object's permissions are `0000`, `osxfs`
```
Comment (mstanleyjones) [inline docker-for-mac/osxfs.md]: Let's try to reduce the complexity of this sentence. I'm thinking something along these lines.

"If the user doesn't have permission to read extended attributes on an object (such as when that object's permissions are `0000`), `osxfs` attempts to add an access control list (ACL) which allows the user to read and write extended attributes. If this attempt fails, the object will appear to be owned by the process accessing it until the extended attribute is readable."

Is it very likely that `osxfs` will be able to add it? I suppose it runs as root?

Comment (samoht) [inline docker-for-mac/osxfs.md]: no it doesn't run as root.

Comment (samoht) [inline docker-for-mac/osxfs.md]: osxfs will have the same privileges that the user running `Docker.app` which basically means the containers will be able to read/write the same sets of files as the user can.

File: docker-for-mac/osxfs.md (hunk 1 above)
Comment (dsheets) [inline docker-for-mac/osxfs.md]: The rest of the file uses "OS X" and "file system". Do you think they should be changed globally? I have no idea what to do about Apple's rename of their OS and the component is still called `osxfs` but maybe the docs should be universally updated... As for "file system", I interpret it as a noun phrase with "file" modifying "system" but I have seen lots of compound word spellings -- what is the convention in the rest of the Docker/Docker for Mac docs?

**Comment by londoncalling** [comment]
@dsheets I incorporated @mstanleyjones 's comments, please merge when reviewed, Misty

File: docker-for-mac/osxfs.md (hunk 2 above)
Comment (dsheets) [inline docker-for-mac/osxfs.md]: I think this is a good simplification, thanks! I endorse Victoria's commit 415dfd7 as accurate.

As for your questions, it is not currently mentioned in the docs but this also applies to attempts to write extended attributes to persist the ownership metadata from the Linux VM. It is very likely that the reason for `chown` failure is that the UNIX permissions on the object prohibit read, write, or both which correspond to reading xattrs, writing xattrs, or both. We take advantage of a couple nice properties of the OS X FS permission model:

1. the owner of a file may change the security settings of that file regardless of the permissions
2. ACLs are included in the security settings
3. ACLs override UNIX permissions

The use case that precipitated this change was the demand for POSIX-compliance regarding the `chown`ing of read-only or mode 0 files (that are still owned by the user). By default, OS X will disallow this but now we can set a bottom priority ACL entry (or add to the first user-allow ACL entry) the ability to read and write extended attributes which then lets us treat extended attributes just like traditional UNIX ownership metadata.

The alternative, without using ACLs, is a horrible non-atomic racy dance of permission juggling as we attempt to evade the coupling of UNIX permissions and xattrs permissions.

I hope this helps elucidate the use case and issues. I'm happy to help improve these docs even more if you can think of ways they could be made clearer.

Thanks!

**Comment by londoncalling** [comment]
Got it @dsheets , made that correction and now will merge this, @mstanleyjones merging now since I think I incorporated all your comments as well.

**Pull request merged**